src/App.tsx

```diff
@@ -1,26 +1,248 @@
+  const [updatedTodosId, setUpdatedTodosId] = useState<number[]>([]);
+  const [failedTitleUpdateId, setFailedTitleUpdateId] = useState<number>(0);
+
+  const downloadTodos = async () => {
+    try {
+      const tmp = await getTodos();
+
+      setTodos(tmp);
+    } catch {
+      setError(ErrorType.UnableToLoad);
+    }
+  };
+
+  useEffect(() => {
+    downloadTodos();
+    inputRef.current?.focus();
+  }, []);
+
+  const filterTodos = () => {
+    if (query === FilterType.Active) {
+      return todos.filter(todo => !todo.completed);
+    } else if (query === FilterType.Completed) {
+      return todos.filter(todo => todo.completed);
+    } else {
+      return todos;
+    }
+  };
+
+  const filteredTodos = filterTodos();
+
+  const returnLeftNumber = () => {
+    return todos.filter(todo => todo.completed != true).length;
+  };
+
+  const handleSetQuery = (passedQuery: FilterType) => {
+    setQuery(passedQuery);
+  };
+
+  const handleSetError = (errorType: ErrorType | null) => {
+    setError(errorType);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+    if (inputRef.current?.value.trim()) {
+      inputRef.current.disabled = true;
+      setTmpTodo({
+        title: inputRef.current.value.trim(),
+        id: -1,
+        completed: false,
+      });
+
+      postTodo(inputRef.current.value.trim())
+        .then(response => {
+          setTodos(prev => [...prev, response]);
+
+          if (inputRef.current) {
+            inputRef.current.value = '';
+          }
+        })
+        .catch(() => setError(ErrorType.UnableToAdd))
+        .finally(() => {
+          if (inputRef.current) {
+            inputRef.current.disabled = false;
+            inputRef.current.focus();
+          }
+
+          setTmpTodo(null);
+        });
+    } else {
+      setError(ErrorType.EmptyTitle);
+    }
+  };
+
+  const handleRemoveTodoById = async (id: number): Promise<boolean> => {
+    setUpdatedTodosId(prev => [...prev, id]);
+    let success = false;
+
+    try {
+      await removeTodo(id);
+
+      setTodos(prev => prev.filter(el => el.id != id));
+      inputRef.current?.focus();
+      success = true;
+    } catch {
+      handleSetError(ErrorType.UnableToDelete);
+    } finally {
+      setUpdatedTodosId(prev => prev.filter(el => el != id));
+    }
+
+    /* removeTodo(id)
+      .then(() => {
+        setTodos(prev => prev.filter(el => el.id != id));
+        success = true;
+      })
+      .catch(() => setError(ErrorType.UnableToDelete))
+      .finally(() => {
+        setUpdatedTodosId(prev => prev.filter(el => el != id));
+        inputRef.current?.focus();
+      });
+    */
+    return success;
+  };
+
+  const handleRemoveAllComplited = () => {
+    const todosToRemoveIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setUpdatedTodosId(todosToRemoveIds);
+
+    todosToRemoveIds.map(id => {
+      handleRemoveTodoById(id);
+    });
```

Comment:
you are maping your todos twice, also why map?
You realy need here only one forEach